qualtran/bloqs/arithmetic/sign_extension.py

```diff
@@ -0,0 +1,125 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+
+import numpy as np
+from attrs import frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    QDType,
+    QFxp,
+    Register,
+    Side,
+    Signature,
+    Soquet,
+    SoquetT,
+)
+from qualtran.bloqs.mcmt import MultiTargetCNOT
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+from qualtran.symbolics import is_symbolic
+
+
+@frozen
+class SignExtend(Bloq):
```

Comment:
Since we need to support the adjoint, should the bloq just be called `Resize` ? It's better to explicitly support resizing down and add tests than expecting `Adjoint(SignExtend())` to do the right thing? 

Also, if we are adding a `Resize` we can explain in docstring the behavior of resizing a signed vs unsigned type and probably just support resizing both instead of supporting resizing only the signed types.

Reply:
Up to the consensus, but my 2 cents would be 1) resize definitely sounds more like a bookkeeping operation and hides the fact that there's a tiny bit of logic to respect the sign bit 2) the linked wiki page is called "sign extend"

Reply:
Ops, I missed the Wikipedia page. SignExtend LGTM then -- let's just support both signed and unsigned types and document the behavior for Adjoint / add a new Bloq for SignTruncate (Truncate seems to be used as the opposite of Extend at some places)? 

https://jarrettbillingsley.github.io/teaching/classes/cs0447/slides/03%20-%20Signed%20Integers,%20Extension,%20Truncation,%20and%20Addition.pptx

Reply:
> let's just support both signed and unsigned types

+1

---

my decomposition supports both directions https://github.com/quantumlib/Qualtran/pull/1163

Reply:
added `SignTruncate` for the adjoint. This PR supports signed values only.

In a follow-up, we can add unsigned support using a separate bloq (say `ZeroExtend` to stick to convention) and a support bloq `ResizeInt` which delegates to the correct extension or truncation bloq based on the dtypes

Reply:
with sign extend and zero extend we will have a pattern of checking dtype before using the appropiate bloq

```py3
if isinstance(self.dtype, ...): bb.add(SignExtend...)
else: bb.add(ZeroExtend...)
```

how about `ArithmeticResize` for a bloq that contains both logics?

Reply:
I think it's fine to merge this PR and add the logic of zero extend either in the same bloq, or a new bloq.